firedrake/slate/slate.py

```diff
@@ -0,0 +1,581 @@
+        return ["S", "V", "M"][self.rank] + "_%d" % self.id
+
+    def __repr__(self):
+        """SLATE representation of the tensor object."""
+        return ["Scalar", "Vector", "Matrix"][self.rank] + "(%r)" % self.form
+
+
+class UnaryOp(TensorBase):
+    """An abstract SLATE class for representing unary operations on a
+    Tensor object.
+
+    The currently supported unary operations on a SLATE :class:`TensorBase`
+    expression are the following:
+
+        (1) the inverse of a tensor, `A.inv`, implemented in the subclass
+            `Inverse`;
+        (2) the transpose of a tensor, `A.T`, implemented in the subclass
+            `Transpose`;
+        (3) and the negative operation, `-A` (subclass `Negative`).
+
+    :arg A: a :class:`TensorBase` object. This can be a terminal tensor object
+            (:class:`Tensor`) or any derived expression resulting from any number
+            of linear algebra operations on `Tensor` objects. For example,
+            another instance of a `UnaryOp` object is an acceptable input, or
+            a `BinaryOp` object.
+    """
+
+    def __init__(self, A):
+        """Constructor for the UnaryOp class."""
+        super(UnaryOp, self).__init__()
+        self.tensor = A
+
+    def coefficients(self):
+        """Returns a tuple of coefficients associated with the tensor."""
+        return self.tensor.coefficients()
+
+    def ufl_domains(self):
+        """Returns the integration domains of the integrals associated with
+        the tensor.
+        """
+        return self.tensor.ufl_domains()
+
+    def subdomain_data(self):
+        """Returns a mapping on the tensor:
+        ``{domain:{integral_type: subdomain_data}}``.
+        """
+        return self.tensor.subdomain_data()
+
+    @property
+    def operands(self):
+        """Returns an iterable of the operands of this operation."""
+        return (self.tensor,)
+
+    def __repr__(self):
+        """SLATE representation of the resulting tensor."""
+        return "%s(%r)" % (type(self).__name__, self.tensor)
+
+
+class Inverse(UnaryOp):
+    """An abstract SLATE class representing the tranpose of a tensor."""
+
+    prec = None
+
+    def __init__(self, A):
+        """Constructor for the Inverse class."""
+        assert A.rank == 2, "The tensor must be rank 2."
+        assert A.shape[0] == A.shape[1], "The inverse can only be computed on square tensors."
+        super(Inverse, self).__init__(A)
+
+    def arguments(self):
+        """Returns the expected arguments of the resulting tensor of
+        performing a specific unary operation on a tensor.
+        """
+        return self.tensor.arguments()[::-1]
+
+    def _output_string(self, prec=None):
+        """Creates a string representation of the inverse of a square tensor."""
+        return "(%s).inv" % self.tensor
+
+    def __str__(self):
+        """Returns a string representation."""
+        return self._output_string(self.prec)
```

Comment:
This function is the same in every class (except in `Tensor`, for no good reason), why is it repeated so many times? You can put this in the base class.